feat: intercept ssh keyManager facade in model proxy

kian99 opened the pull request:
## Description
Currently this PR builds on #1521.

This PR allows JIMM to intercept model calls for the SSH KeyManager facade and direct the calls to JIMM's business logic rather than forwarding the calls to the backing Juju controllers.

This is being done in order to support the SSH proxy efforts.
Ideally, in Juju 4 these methods would be done on the controller api and all this logic would move into the jujuapi package. I've left comments to that effect. If before that point we want to refactor the model proxy then the logic has been cleanly split so that the code you see inside `internal/rpcproxy/keymanager.go` is all the logic/code that would go inside of `jujuapi`.

I've manually tested the changes by spinning up JIMM and adding a Juju 3.5 controller and tried the following commands (not in this specific order):
```
juju list-ssh-keys
juju list-ssh-keys --full
juju add-ssh-key "$(cat somekey.pub)"
juju remove-ssh-key <comment>
juju remove-ssh-key <fingerprint>
```

Fixes [JUJU-7346](https://warthogs.atlassian.net/browse/JUJU-7346)

## Engineering checklist
<!-- *Check only items that apply* -->

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

[JUJU-7346]: https://warthogs.atlassian.net/browse/JUJU-7346?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kian99 committed:
feat: add keyManager handler to model proxy

The model proxy will intercept calls to the keyManager facade and persist user keys in JIMM rather than passing these calls along to the Juju controller. This is being done in order to support the SSH proxy efforts.
Ideally, in Juju 4 these methods would be done on the controller api and all this logic would move into the jujuapi package.

alesstimec commented on internal/rpcproxy/rpcproxy.go:
```diff
@@ -710,3 +746,63 @@ func (p *clientProxy) handleAdminFacade(ctx context.Context, msg *message) (clie
+// handleKeyManagerFacade processes the key manager facade call.
+func (p *clientProxy) handleKeyManagerFacade(ctx context.Context, msg *message) (clientResponse *message, err error) {
+	if p.user == nil {
+		return nil, errors.E("user not authenticated")
```
should we not return the error in `clientResponse`? i.e. fill in the `Error` and `ErrorCode` fields.

kian99 replied:
That gets handled outside this function

alesstimec commented on internal/rpcproxy/sshkeys.go:
```diff
@@ -0,0 +1,123 @@
+	user *openfga.User
+}
+
+func (s *keyManagerFacade) ListKeys(ctx context.Context, args jujuparams.ListSSHKeys) (jujuparams.StringsResults, error) {
```
godoc, please

alesstimec commented on internal/rpcproxy/sshkeys.go:
```diff
@@ -0,0 +1,123 @@
+		return jujuparams.StringsResults{}, err
+	}
+
+	var formatter func(key sshkeys.PublicKey) string
```
is this what juju does? it seems super suspicious.. i would expect formatting to happen in the cmd package

kian99 replied:
Yes it's all done server side - https://github.com/juju/juju/blob/3.6/apiserver/facades/client/keymanager/keymanager.go#L111
It seems to be like this because of legacy reasons, i.e. the return type is a list of strings rather than some well defined type that the client can then format.

alesstimec commented on internal/rpcproxy/sshkeys.go:
```diff
@@ -0,0 +1,123 @@
+	return jujuparams.StringsResults{Results: []jujuparams.StringsResult{res}}, nil
+}
+
+func (s *keyManagerFacade) AddKeys(ctx context.Context, args jujuparams.ModifyUserSSHKeys) (jujuparams.ErrorResults, error) {
```
godoc, please

alesstimec commented on internal/rpcproxy/sshkeys.go:
```diff
@@ -0,0 +1,123 @@
+// Copyright 2025 Canonical.
```
could we rename this file to keymanagerfacade.go?

kian99 replied:
What about just `keymanager.go`? Changing it to that for now.

alesstimec commented on internal/rpcproxy/sshkeys.go:
```diff
@@ -0,0 +1,123 @@
+	return jujuparams.ErrorResults{Results: res}, nil
+}
+
+func (s *keyManagerFacade) DeleteKeys(ctx context.Context, args jujuparams.ModifyUserSSHKeys) (jujuparams.ErrorResults, error) {
```
godoc, please

kian99 committed:
chore: rename sshkeys and add godocs

alesstimec commented on internal/rpcproxy/keymanager.go:
```diff
@@ -0,0 +1,129 @@
+// with the authenticated user.
+func (s *keyManagerFacade) AddKeys(ctx context.Context, args jujuparams.ModifyUserSSHKeys) (jujuparams.ErrorResults, error) {
+	var res []jujuparams.ErrorResult
+	errF := func(err error, msg string) jujuparams.ErrorResult {
```
suggestion: just have `func(msg string) jujuparams.ErrorResult`.. you are already creating a message string every time you call `errF`, so appending the `err.Error()` each time, would be a minor change.

kian99 replied:
Good point, as I was changing it I noticed `jujuparams.Error` also accepts an error code. If I want to fill that in I need to keep the error in the signature. I'm now populating the error code, so I've mostly left things the same, let me know what you think.

alesstimec replied:
then perhaps 
```
func(msg, errorCode string) jujuparams.ErrorResult {
  		return jujuparams.ErrorResult{Error: &jujuparams.Error{
			Code:    string(errors.ErrorCode(err)),
			Message: msg,
		}}
}
```
but just a suggestion, i won't insist on it..

kian99 replied:
It's minor so I've left it as is.

alesstimec commented on internal/rpcproxy/keymanager.go:
```diff
@@ -0,0 +1,129 @@
+// and finds keys to remove either by comment or fingerprint.
+func (s *keyManagerFacade) DeleteKeys(ctx context.Context, args jujuparams.ModifyUserSSHKeys) (jujuparams.ErrorResults, error) {
+	var res []jujuparams.ErrorResult
+	errF := func(err error, msg string) jujuparams.ErrorResult {
```
same suggestion

kian99 committed:
chore: populate error code

ale8k commented on internal/rpcproxy/keymanager.go:
```diff
@@ -0,0 +1,132 @@
+// the api layer logic for SSH key management methods that are currently
+// used by the rpcProxy.
+type keyManagerFacade struct {
+	SSHKeyManager
```
Why is this embedded field? I often prefer named fields so it's really clear what struct we're calling methods on.

kian99 replied:
Good catch, I'm not sure either, it's not necessary to be embedded, I've changed it.

ale8k commented on internal/rpcproxy/keymanager.go:
```diff
@@ -0,0 +1,132 @@
+		if err != nil {
+			res = append(res, errF(err, fmt.Sprintf("Failed to parse key (entry %d)", i)))
+		}
+		jimmKey := sshkeys.PublicKey{
```
parsedKey* maybe a better name?

ale8k commented on internal/rpcproxy/keymanager.go:
```diff
@@ -0,0 +1,132 @@
+
+func marshalAuthorizedKeyWithComment(key sshkeys.PublicKey) string {
+	// Copied from gossh.MarshalAuthorizedKey with an addition for the comment.
+	// Errors from the buffer's Write..() methods are always nil.
```
Move this up to the actual godoc please, also I've forogtten - what is the comment for?

kian99 replied:
An SSH key doesn't have a comment, but the OpenSSH authorized key format for public keys does and Juju stores the key comment and displays it back to the user when they list keys, helping to identify for a user where the key came from.

ale8k commented on internal/rpcproxy/keymanager.go:
```diff
@@ -0,0 +1,132 @@
+	return b.String()
+}
+
+func fingerprintWithComment(key sshkeys.PublicKey) string {
```
Godoc again explaining why we need a comment would be good

ale8k commented on internal/rpcproxy/rpcproxy_test.go:
```diff
@@ -272,3 +280,139 @@ func TestProxySocketsAuditLogs(t *testing.T) {
 
 }
+
+func TestProxySocketsSSHKeys(t *testing.T) {
```
A godoc for this test would be useful

ale8k reviewed (approved):
lgtm some minor

kian99 committed:
chore: extra godocs and added test case

kian99 committed:
chore: add api layer test

This test is a "real" test in that it uses the Juju API client to make a call to add/list/remove a key.

kian99 committed:
chore: don't embed keyManager field
